README.rst
atiiaftt - ATI-IA force-torque transforms
===============================================================

Python interface to ATI-IA force-torque transform (ATIDAQ) C library, v1.0.7. For use with ATI Industrial Automation force-torque sensors.

This package does not read from hardware. Another package, such as NI-DAQmx, 
must be used for that purpose.

ATIDAQ C Library is available at https://www.ati-ia.com/library/software/daq_ft/ATIDAQ%20C%20Library.zip

Installation
------------------------

pip install atiiaftt

Dependencies
------------------------

CFFI and platform appropriate build tools are required to build the C source
that is included with the package. 

For Linux distributions, install the developer package group.

For Windows, see https://wiki.python.org/moin/WindowsCompilers

Documentation
------------------------

Module help for the python classes is available via docstrings. Documentation
for the wrapped functions is available in `ATIDAQ_c_lib/readme.txt`


Usage examples
------------------------

High level usage is as follows:

1. `import atiiaftt`
2. Create an instance of the `atiiafft.FTsensor` class
3. Load calibration data, either passing the calibration file path as a string during class instantiation or as a parameter while calling `instance.createCalibration('./FT18766cal')`.
4. Optionally, set the tool transform, bias values and units as needed, eg)
	- `instance.setToolTransform([0,0,20,45,0,0],atiiaftt.FTUnit.DIST_MM,atiiaftt.FTUnit.ANGLE_DEG)`
	- `instance.bias([0.254,-1.027,0.025,0.7422.0.9302,-0.230,0.082])`
5. Call the force-torque conversion function: `instance.convertToFt([0.042,1.004,0.952,-0.235,0.091,1.091,0.054])`; this returns the forces and torques as a list `[F.x,F.y,F.z,T.x,T.y,T.z]`. The last transformed values are stored in the instance variable `instance.ft_vector` for later access.

CFFI and python class examples found in 'atiiaftt/samples/ftconvert.py'


changelog.txt
# atiiaftt (ATI AI force-torque transforms) changelog

## 0.1.1

2018-10-04444

### Added

- cffi build to MANIFEST.in

2018-10-03

### Added

- cffi build and setup files
- python wrapper/interface classes
- example code for python classes and cffi objects
- unit tests as per `ATIDAQ_c_lib/Samples/ftconvert.c`


atiiaftt/test/test_ft_iface.py
from __future__ import print_function

import os
import sys
import datetime
import logging 

from _atiiaftt import ffi,lib
import atiiaftt 

import unittest 

general_log_formatter = logging.Formatter('%(asctime)s:%(levelname)s:%(name)s:%(message)s')
logging.basicConfig()
logger = logging.getLogger(__name__)
logger.level = logging.DEBUG
log_filepath = os.path.join(".",datetime.datetime.now().strftime('%Y%m%d%H%M')+".atiiaftt.test.log")
#file_log_hndl=logging.FileHandler(log_filepath)
#file_log_hndl.setFormatter(general_log_formatter)
#logger.addHandler(file_log_hndl)

class TestAtiiaFTT(unittest.TestCase):
    def setUp(self):
        self.cal_file="atiiaftt/test/FT18766.cal"
        self.bias_readings=[0.265100,-0.017700,-0.038400,-0.042700,-0.189100,0.137300,-3.242300]
        self.input_floats=[-3.286300,0.387500,-3.487700,0.404300,-3.934100,0.547400,-3.210600]
        self.translation_floats=[0,0,20,45,0,0]
        self.output_floats_n=[-4.285730,-908.400635,-928.162720,-0.679282,-1.102992,-1.988170]
        self.output_floats_kn=[-0.004286, -0.908401, -0.928163, -0.679282, -1.102992, -1.98817]
        self.not_a_unit=ffi.new("char[]","invalid".encode("ascii"))
        self.test_sensor=atiiaftt.FTSensor()

    def test_FTSensor_CreateCal_OK(self):
        """
        Test that a calibration file can be loaded.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.assertNotEqual(self.test_sensor.calibration,ffi.NULL)

    def test_FTSensor_CreateCal_FAIL01(self):
        """
        Test that an exception is raised when the calibration index is out of bounds.
        """
        self.assertRaises(IndexError,self.test_sensor.createCalibration,self.cal_file,0)

    def test_FTSensor_CreateCal_FAIL02(self):
        """
        Test that an exception is rased when the file path does not point to a calibration file.
        """
        self.assertRaises(IOError,self.test_sensor.createCalibration,"/not-a-calfile.txt",1)

    def test_FTSensor_Bias_OK01(self):
        """
        Test that a bias vector can be stored.
        """
        logger.info("start test 'test_FTSensor_Bias01'")
        self.test_sensor.createCalibration(self.cal_file,1)
        logger.info("Created {}".format(self.test_sensor))
        logger.debug("Calibration Object: {}".format(self.test_sensor.calibration))
        self.test_sensor.bias(self.bias_readings)
        logger.debug(self.test_sensor.bias)
        self.assertEqual(self.bias_readings,self.test_sensor.bias_vector)

    def test_FTSensor_Bias_OK02(self):
        """
        Test that a bias vector can be stored by checking the cdata values.
        """
        logger.info("start test 'test_FTSensor_Bias02'")
        self.test_sensor.createCalibration(self.cal_file,1)
        logger.info("Created {}".format(self.test_sensor))
        logger.debug("Calibration Object: {}".format(self.test_sensor.calibration))
        self.test_sensor.bias(self.bias_readings)
        cffi_bias_vector=self.test_sensor.calibration.rt.bias_vector
        logger.debug(cffi_bias_vector)
        bias_vector=[]
        for c_bias in cffi_bias_vector:
            logger.debug("{}".format(c_bias))
            if c_bias!=0.0:
                bias_vector.append(round(c_bias,6))
        logger.debug("{}".format(bias_vector))
        self.assertEqual(self.bias_readings,bias_vector)

    def test_FTSensor_Bias_FAIL01(self):
        """
        Test that an exception is raised while trying to bias an instance with no calibration.
        """
        #self.test_sensor.createCalibration(self.cal_file,1)
        #self.test_sensor.bias(self.bias_readings)
        self.assertRaises(RuntimeError,self.test_sensor.bias,self.bias_readings)

    def test_FTSensor_SetTT_OK01(self):
        """
        Test the setToolTransform wrapper function.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setToolTransform(self.translation_floats,atiiaftt.FTUnit.DIST_MM,atiiaftt.FTUnit.ANGLE_DEG)
        new_aunit=ffi.string(self.test_sensor.calibration.cfg.UserTransform.AngleUnits)
        new_dunit=ffi.string(self.test_sensor.calibration.cfg.UserTransform.DistUnits)
        tt_vector=list(self.test_sensor.calibration.cfg.UserTransform.TT)
        logger.debug("{}, {}".format(new_aunit,new_dunit))
        logger.debug("{}".format(tt_vector))
        self.assertEqual(b"deg",new_aunit)
        self.assertEqual(b"mm",new_dunit)
        self.assertEqual(self.translation_floats,tt_vector)

    def test_FTSensor_SetTT_OK02(self):
        """
        Test the setToolTransform wrapper function.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setToolTransform(self.translation_floats,atiiaftt.FTUnit.DIST_FT,atiiaftt.FTUnit.ANGLE_RAD)
        new_aunit=ffi.string(self.test_sensor.calibration.cfg.UserTransform.AngleUnits)
        new_dunit=ffi.string(self.test_sensor.calibration.cfg.UserTransform.DistUnits)
        tt_vector=list(self.test_sensor.calibration.cfg.UserTransform.TT)
        logger.debug("{}, {}".format(new_aunit,new_dunit))
        logger.debug("{}".format(tt_vector))
        self.assertEqual(b"rad",new_aunit)
        self.assertEqual(b"ft",new_dunit)
        self.assertEqual(self.translation_floats,tt_vector)
 
    def test_FTSensor_SetTT_FAIL01(self):
        """
        Test that an exception is raised while trying to use the tool transform on an instance with no calibration.
        """
        self.assertRaises(RuntimeError,self.test_sensor.setToolTransform,self.translation_floats,atiiaftt.FTUnit.DIST_MM,atiiaftt.FTUnit.ANGLE_DEG)
 
    def test_FTSensor_SetTT_FAIL02(self):
        """
        Test that an exception is raised while trying to using an invalid unit with the tool transform wrapper function.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.assertRaises(ValueError,self.test_sensor.setToolTransform,self.translation_floats,self.not_a_unit,atiiaftt.FTUnit.ANGLE_DEG)
 
    def test_FTSensor_SetTT_FAIL03(self):
        """
        Test that an exception is raised while trying to using an invalid unit with the tool transform wrapper function.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.assertRaises(ValueError,self.test_sensor.setToolTransform,self.translation_floats,atiiaftt.FTUnit.DIST_MM,self.not_a_unit)
 
    def test_FTSensor_SetForceUnits_OK01(self):
        """
        Test that the setForceUnits wrapper function properly sets 'FORCE_LB' units.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_LB)
        new_funit=ffi.string(self.test_sensor.calibration.cfg.ForceUnits)
        self.assertEqual(b"lb",new_funit)

    def test_FTSensor_SetForceUnits_OK02(self):
        """
        Test that the setForceUnits wrapper function properly sets 'FORCE_KG' units.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_KG)
        new_funit=ffi.string(self.test_sensor.calibration.cfg.ForceUnits)
        self.assertEqual(b"kg",new_funit)

    def test_FTSensor_SetForceUnits_OK03(self):
        """
        Test that the setForceUnits wrapper function properly sets 'FORCE_KN' units.
        """
        self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_KN)
        new_funit=ffi.string(self.test_sensor.calibration.cfg.ForceUnits)
        self.assertEqual(b"kN",new_funit)

    def test_FTSensor_SetForceUnits_OK04(self):
        """
        Test that the setForceUnits wrapper function properly sets 'FORCE_G' units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_G)
        new_funit=ffi.string(self.test_sensor.calibration.cfg.ForceUnits)
        self.assertEqual(b"g",new_funit)

    def test_FTSensor_SetForceUnits_OK06(self):
        """
        Test that the setForceUnits wrapper function properly sets 'FORCE_KLB' units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_KLB)
        new_funit=ffi.string(self.test_sensor.calibration.cfg.ForceUnits)
        self.assertEqual(b"klb",new_funit)

    def test_FTSensor_SetForceUnits_FAIL01(self):
        """
        Test that an exception is raised when using the setForceUnits wrapper function 
        on an instance with no calibration data.
        """
        self.assertRaises(RuntimeError,self.test_sensor.setForceUnits,atiiaftt.FTUnit.DIST_MM)

    def test_FTSensor_SetForceUnits_FAIL02(self):
        """
        Test that an exception is raised when using the setForceUnits wrapper function 
        on an instance with an incorrect unit.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.assertRaises(ValueError,self.test_sensor.setForceUnits,self.not_a_unit)

    def test_FTSensor_SetTorqueUnits_OK01(self):
        """
        Test that the setTorqueUnits wrapper function properly sets 'TORQUE_IN_LB' units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setTorqueUnits(atiiaftt.FTUnit.TORQUE_IN_LB)
        new_tunit=ffi.string(self.test_sensor.calibration.cfg.TorqueUnits)
        self.assertEqual(b"in-lb",new_tunit)

    def test_FTSensor_SetTorqueUnits_OK02(self):
        """
        Test that the setTorqueUnits wrapper function properly sets 'TORQUE_N_M' units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setTorqueUnits(atiiaftt.FTUnit.TORQUE_N_M)
        new_tunit=ffi.string(self.test_sensor.calibration.cfg.TorqueUnits)
        self.assertEqual(b"N-m",new_tunit)

    def test_FTSensor_SetTorqueUnits_OK03(self):
        """
        Test that the setTorqueUnits wrapper function properly sets 'TORQUE_N_MM' units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setTorqueUnits(atiiaftt.FTUnit.TORQUE_N_MM)
        new_tunit=ffi.string(self.test_sensor.calibration.cfg.TorqueUnits)
        self.assertEqual(b"N-mm",new_tunit)

    def test_FTSensor_SetTorqueUnits_OK04(self):
        """
        Test that the setTorqueUnits wrapper function properly sets 'TORQUE_KG_CM' units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.setTorqueUnits(atiiaftt.FTUnit.TORQUE_KG_CM)
        new_tunit=ffi.string(self.test_sensor.calibration.cfg.TorqueUnits)
        self.assertEqual(b"kg-cm",new_tunit)

    def test_FTSensor_SetTorqueUnits_FAIL01(self):
        """
        Test that an exception is raised when using the setTorqueUnits wrapper function 
        on an instance with no calibration data.
        """
        self.assertRaises(RuntimeError,self.test_sensor.setTorqueUnits,atiiaftt.FTUnit.TORQUE_IN_LB)

    def test_FTSensor_SetTorqueUnits_FAIL02(self):
        """
        Test that an exception is raised when using the setTorqueUnits wrapper function 
        on an instance with an incorrect unit.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.assertRaises(ValueError,self.test_sensor.setTorqueUnits,self.not_a_unit)

    def test_FTSensor_ConvertToFt_OK01(self):
        """
        Test the convertToFt wrapper function with a specific tool transform and units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.bias(self.bias_readings)
        self.test_sensor.setToolTransform(self.translation_floats,atiiaftt.FTUnit.DIST_MM,atiiaftt.FTUnit.ANGLE_DEG)
        self.test_sensor.setTorqueUnits(atiiaftt.FTUnit.TORQUE_N_M)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_N)
        self.test_sensor.convertToFt(self.input_floats)
        logger.debug(self.test_sensor.ft_vector)
        rnd_ft_vector=[]
        for c_ft in self.test_sensor.ft_vector:
            rnd_ft_vector.append(round(c_ft,6))
        logger.debug(rnd_ft_vector)
        self.assertEqual(self.output_floats_n,rnd_ft_vector)

    def test_FTSensor_ConvertToFt_OK02(self):
        """
        Test the convertToFt wrapper function with a specific tool transform and units.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.bias(self.bias_readings)
        self.test_sensor.setToolTransform(self.translation_floats,atiiaftt.FTUnit.DIST_MM,atiiaftt.FTUnit.ANGLE_DEG)
        self.test_sensor.setTorqueUnits(atiiaftt.FTUnit.TORQUE_N_M)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_KN)
        self.test_sensor.convertToFt(self.input_floats)
        rnd_ft_vector=[]
        for c_ft in self.test_sensor.ft_vector:
            rnd_ft_vector.append(round(c_ft,6))
        logger.debug(rnd_ft_vector)
        self.assertEqual(self.output_floats_kn,rnd_ft_vector)

    def test_FTSensor_ConvertToFt_OK03(self):
        """
        Test the convertToFt wrapper function stores the last conversion values.
        """
        cal_ok=self.test_sensor.createCalibration(self.cal_file,1)
        self.test_sensor.bias(self.bias_readings)
        self.test_sensor.setToolTransform(self.translation_floats,atiiaftt.FTUnit.DIST_MM,atiiaftt.FTUnit.ANGLE_DEG)
        self.test_sensor.setTorqueUnits(atiiaftt.FTUnit.TORQUE_N_M)
        self.test_sensor.setForceUnits(atiiaftt.FTUnit.FORCE_N)
        ft_readings=self.test_sensor.convertToFt(self.input_floats)
        self.assertEqual(self.test_sensor.ft_vector,ft_readings)

    def test_FTSensor_ConvertToFt_FAIL01(self):
        """
        Test that an exception is raised when the convertToFt wrapper function is used on an instance with
        no calibration data.
        """
        self.assertRaises(RuntimeError,self.test_sensor.convertToFt,self.input_floats)


if __name__ == "__main__":

    unittest.main()


cffi_build.py
import os
from cffi import FFI

ffibuilder = FFI()

# have this build script with the ATIDAQ source for now
PATH = os.path.dirname(__file__)

with open(os.path.join(PATH, 'ATIDAQ_c_lib/ATIDAQ/ftconfig.h'),'r') as f:
    ffibuilder.set_source("_atiiaftt", f.read(),
                          sources=[os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/ftconfig.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/ftrt.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/dom.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/expatls.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/node.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/stack.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/xmlparse.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/xmlrole.c'),
                                   os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ/xmltok.c')
                                   ],
                          include_dirs=[os.path.join(PATH,'ATIDAQ_c_lib/ATIDAQ')]
                          )

ffibuilder.cdef(
    """
    #define MAX_AXES 6
    #define MAX_GAUGES 8

    typedef char *Units;
    typedef struct Configuration Configuration;
    typedef struct Calibration Calibration;
    typedef struct Transform Transform;
    typedef int BOOL;

    typedef struct RTCoefs RTCoefs;

    struct RTCoefs {
        unsigned short NumChannels;
        unsigned short NumAxes;
        float working_matrix[MAX_AXES][MAX_GAUGES];
        float bias_slopes[MAX_GAUGES];
        float gain_slopes[MAX_GAUGES];
        float thermistor;
        float bias_vector[MAX_GAUGES+1];
        float TCbias_vector[MAX_GAUGES];
    };

    struct Transform {
        float TT[6];
        Units DistUnits;
        Units AngleUnits;
    };


    struct Configuration {
        Units ForceUnits;        // force units of output
        Units TorqueUnits;       // torque units of output
        Transform UserTransform; // coordinate system transform set by user
        BOOL TempCompEnabled;    // is temperature compensation enabled?
    };

    struct Calibration {
        float BasicMatrix[MAX_AXES][MAX_GAUGES];
        Units ForceUnits;
        Units TorqueUnits;
        BOOL TempCompAvailable;
        Transform BasicTransform;
        float MaxLoads[MAX_AXES];
        char *AxisNames[MAX_AXES];
        char *Serial;
        char *BodyStyle;
        char *PartNumber;
        char *Family;
        char *CalDate;
        Configuration cfg;
        RTCoefs rt;
    };

    Calibration *createCalibration(char *CalFilePath, unsigned short index);
    void destroyCalibration(Calibration *cal);
    short SetToolTransform(Calibration *cal, float Vector[6],char *DistUnits,char *AngleUnits);
    short SetForceUnits(Calibration *cal, char *NewUnits);
    short SetTorqueUnits(Calibration *cal, char *NewUnits);
    void Bias(Calibration *cal, float voltages[]);
    void ConvertToFT(Calibration *cal, float voltages[],float result[]);

    """
    )

if __name__ == "__main__":
    ffibuilder.compile(verbose=True)


atiiaftt/__init__.py
"""
Force-Torque transform library for ATI-IA FT Sensors.

This module uses cffi to wrap the MIT licensed c library released by
ATI-IA for force-torque sensors.

The tranform functions may be called by importing the cffi module and 
calling the application interfaces directly, or the wrapper classes 
may be used to call the functions with more pythonic methods.

Usage examples are found in 'ftconvert.py'

This module does not provide any access to hardware, a module such as 
'NI-DAQmx', matching the ADC, must be used to aquire data first.

Copyright (c) 2018 Tyson Boer

The MIT License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

name = "atiiaftt"

__all__ = ['FTUnit','FTSensor']

from .ft_interface import FTUnit
from .ft_interface import FTSensor



atiiaftt/ft_interface.py
import os
import sys
import logging
from _atiiaftt import ffi,lib

class FTUnit:
    """
    Convenience class holding accepted units for the atiia functions. 
    
    Constants declared from 'ftconfig.c'.
    """
    FORCE_LB=ffi.new("char[]","lb".encode("ascii"))
    """lb\\0"""
    FORCE_KLB=ffi.new("char[]","klb".encode("ascii"))
    """klb\\0"""
    FORCE_N=ffi.new("char[]","N".encode("ascii"))
    """N\\0"""
    FORCE_KN=ffi.new("char[]","kN".encode("ascii"))
    """kN\\0"""
    FORCE_G=ffi.new("char[]","g".encode("ascii"))
    """g\\0"""
    FORCE_KG=ffi.new("char[]","kg".encode("ascii"))
    """kg\\0"""
    TORQUE_IN_LB=ffi.new("char[]","in-lb".encode("ascii"))
    """in-lb\\0"""
    TORQUE_FT_LB=ffi.new("char[]","ft-lb".encode("ascii"))
    """ft-lb\\0"""
    TORQUE_N_M=ffi.new("char[]","N-m".encode("ascii"))
    """N-m\\0"""
    TORQUE_N_MM=ffi.new("char[]","N-mm".encode("ascii"))
    """N-mm\\0"""
    TORQUE_KG_CM=ffi.new("char[]","kg-cm".encode("ascii"))
    """kg-cm\\0"""
    DIST_M=ffi.new("char[]","m".encode("ascii"))
    """m\\0"""
    DIST_CM=ffi.new("char[]","cm".encode("ascii"))
    """cm\\0"""
    DIST_MM=ffi.new("char[]","mm".encode("ascii"))
    """mm\\0"""
    DIST_FT=ffi.new("char[]","ft".encode("ascii"))
    """ft\\0"""
    ANGLE_DEG=ffi.new("char[]","deg".encode("ascii"))
    """deg\\0"""
    ANGLE_RAD=ffi.new("char[]","rad".encode("ascii"))
    """rad\\0"""

class FTSensor:
    """
    Wrapper class to hold force-torque sensor calibration and values.
    """
    def __init__(self,CalFilePath=None,index=1):
        """
        Force-torque sensor instance constructor. Optionally takes calibration file info.

        @param CalFilePath: Path string passed to self.createCalibration
        @type CalFilePath: string
        @param index: Index value passed to self.createCalibration
        @type index: int
        """
        self.logger=logging.getLogger("atiiaftt.FTSensor")

        self.calibration=ffi.NULL
        if CalFilePath!=None:
            self.createCalibration(CalFilePath,index)

        self.voltage_vector=[]
        self.bias_vector=[]
        self.ft_vector=[]

    def createCalibration(self,CalFilePath,index):
        """
        Wraps c function 'createCalibration()'.

        @param CalFilePath: Path string of calibration file for the sensor
        @type CalFilePath: string
        @param index: Index value of the requested calibration data.
        @type index: int
        @raises IOError: exception raised if 'CalFilePath' fails os.path.exists()
        @raises IndexError: exception raised if index value not found in calibration file
        """
        if not (os.path.exists(CalFilePath)):
            self.logger.error("Can't find calibration file: "+CalFilePath)
            raise IOError("Calibration file not found.")

        cffi_cal_filepath=ffi.new("char[]",CalFilePath.encode("ascii"))
        self.calibration=lib.createCalibration(cffi_cal_filepath,index)
        if self.calibration==ffi.NULL:
            self.logger.error("Index not found in calibration file.")
            raise IndexError("Passed calibration index not found.")
        # else ok

    def setToolTransform(self,vector,distunits,angleunits):
        """
        Wraps c function 'SetToolTransform()'.

        @param vector: 
        @type vector: list of floats
        @param distunits:
        @type distunits: atiiaftt.FTUnit.DIST_* class member
        @param angleunits:
        @type angleunits: atiiaftt.FTUnit.ANGLE_* class member
        @raises RuntimeError: exception raised if function is called before loading a calibration dataset.
        @raises ValueError: exception raised if unknown unit is passed.
        """
        cffi_vector=ffi.new("float[]",vector)
        ret_val=lib.SetToolTransform(self.calibration,cffi_vector,distunits,angleunits)
        if ret_val == 1:
            raise RuntimeError("Calibration data not loaded.")
        elif ret_val == 2:
            raise ValueError("Invalid distance unit.")
        elif ret_val == 3:
            raise ValueError("Invalid angle unit.")

    def setForceUnits(self,newunits):
        """
        Wraps c function 'SetForceUnits()'

        @param newunits: New force unit to used. Stored in 'calibration.cfg.ForceUnits'
        @type newunits: atiiaftt.FTUnit.FORCE_* class member
        @raises RuntimeError: exception raised if function is called before loading a calibration dataset.
        @raises ValueError: exception raised if unknown unit is passed.

        """
        ret_val=lib.SetForceUnits(self.calibration,newunits)
        if ret_val == 1:
            raise RuntimeError("Calibration data not loaded.")
        elif ret_val == 2:
            raise ValueError("Invalid force unit.")

    def setTorqueUnits(self,newunits):
        """
        Wraps c function 'SetTorqueUnits()'

        @param newunits: New torque unit to used. Stored in 'calibration.cfg.TorqueUnits'
        @type newunits: atiiaftt.FTUnit.TORQUE_* class member
        @raises RuntimeError: Exception raised if function is called before loading a calibration dataset.
        @raises ValueError: Exception raised if unknown unit is passed.

        """
        ret_val=lib.SetTorqueUnits(self.calibration,newunits)
        if ret_val == 1:
            raise RuntimeError("Calibration data not loaded.")
        elif ret_val == 2:
            raise ValueError("Invalid torque unit.")

    def bias(self,voltages):
        """
        Wraps c function 'Bias()'

        @param voltages: Averaged set of ADC readings from a sensor
        @type voltages: list of floats, also stored in self.bias_vector
        @raises RuntimeError: exception raised if function is called before loading a calibration dataset.

        """
        self.bias_vector=voltages
        if self.calibration==ffi.NULL:
            self.logger.error("Calibration data not loaded.")
            raise RuntimeError("Calibration data not loaded.")

        cffi_bias=ffi.new("float[]",voltages)
        self.logger.debug("{}".format(cffi_bias))
        lib.Bias(self.calibration,cffi_bias)

    def convertToFt(self,voltages):
        """
        Wraps c function 'ConvertToFT()'. 
        The most recent voltages are also stored in self.voltage_vector
        The most recent conversion is also stored in self.ft_vector

        @param voltages: values read from an ADC connected to a sensor
        @type voltages: list of floats
        @raises RuntimeError: exception raised if function is called before loading a calibration dataset.
        @return: Force-torque values, also stored in self.ft_vector
        @rtype: list of floats, format [F.x,F.y,F.z,T.x,T.y,T.z]

        """
        self.voltage_vector=voltages
        if self.calibration==ffi.NULL:
            self.logger.error("Calibration data not loaded.")
            raise RuntimeError("Calibration data not loaded.")

        cffi_ft_vector=ffi.new("float[]",[0,0,0,0,0,0])
        cffi_voltages=ffi.new("float[]",voltages)
        lib.ConvertToFT(self.calibration,cffi_voltages,cffi_ft_vector)
        self.ft_vector=list(cffi_ft_vector)
        return self.ft_vector


atiiaftt/samples/ftconvert.py
from __future__ import print_function

import os
import sys
import datetime
import logging 

from _atiiaftt import ffi,lib
import atiiaftt 

def usage_example():
    """
    Usage example for atiiaftt python classes.

    Functionally the same as the Sample/ftconvert.c example included with ATIDAQ zip
    """
    cal_file="tests/FT18766.cal"
    cal_file_index=1

    # same values as sample program 'ftconvert.c'
    bias_readings=[0.265100,-0.017700,-0.038400,-0.042700,-0.189100,0.137300,-3.242300]
    input_floats=[-3.286300,0.387500,-3.487700,0.404300,-3.934100,0.547400,-3.210600]
    translation_floats=[0,0,20,45,0,0]
    output_floats=[0,0,0,0,0,0,0]
    translation_dist_unit=atiiaftt.FTUnit.DIST_MM
    translation_angle_unit=atiiaftt.FTUnit.ANGLE_DEG
    force_unit_str=atiiaftt.FTUnit.FORCE_N
    torque_unit_str=atiiaftt.FTUnit.TORQUE_N_M

    ftSensor00=atiiaftt.FTSensor()
    try:
        forcetorqueValues=ftSensor00.convertToFt(input_floats)
        forcetorqueValues=ftSensor00.bias(bias_readings)
    except RuntimeError:
        logging.error("Caught expected exception.") 

    print("Using calibration file: '{}'".format(cal_file))
    ftSensor01=atiiaftt.FTSensor(cal_file,cal_file_index)
    print(ftSensor01.calibration)

    # print the working calibration matrix (loaded from file, pre-bias)
    print("Calibration Matrix Dimensions [{} Channels,{} Axes]".format(ftSensor01.calibration.rt.NumChannels,ftSensor01.calibration.rt.NumAxes))
    for ai in range(ftSensor01.calibration.rt.NumAxes):
        print("{}\t".format(ffi.string(ftSensor01.calibration.AxisNames[ai])),end="")
        for ci in range(ftSensor01.calibration.rt.NumChannels):
            print("{}\t".format(ftSensor01.calibration.rt.working_matrix[ai][ci]),end="")
        print("")
            
    print("Running tool transform with translation: {}".format(translation_floats))

    ftSensor01.setToolTransform(translation_floats,translation_dist_unit,translation_angle_unit)
    ftSensor01.bias(bias_readings)
    forcetorqueValues=ftSensor01.convertToFt(input_floats)

    print("Bias reading: {}".format(bias_readings))
    print("Measurement: {}".format(input_floats))
    print("Result: {}".format(forcetorqueValues))
    print("Stored Result: {}".format(ftSensor01.ft_vector))
    
    return 0

def cffi_usage_example():
    """
    Usage example for atiiaftt cffi library.

    Functionally the same as the Sample/ftconvert.c example included with ATIDAQ zip
    """
    cal_file="tests/FT18766.cal"
    cal_file_index=1
    cal_file_path=ffi.new("char[]",os.path.join(os.path.dirname(__file__),cal_file).encode("ascii"))

    print(ffi.string(cal_file_path))
    
    calibrationData=lib.createCalibration(cal_file_path,cal_file_index)
    print(calibrationData)

    # print the working calibration matrix (loaded from file, pre-bias)
    print("Calibration Matrix Dimensions [{} Channels,{} Axes]".format(calibrationData.rt.NumChannels,calibrationData.rt.NumAxes))
    for ai in range(calibrationData.rt.NumAxes):
        print("{}\t".format(ffi.string(calibrationData.AxisNames[ai])),end="")
        for ci in range(calibrationData.rt.NumChannels):
            print("{}\t".format(calibrationData.rt.working_matrix[ai][ci]),end="")
        print("")
            
    # same values as sample program 'ftconvert'
    bias_readings=ffi.new("float[]",[0.265100,-0.017700,-0.038400,-0.042700,-0.189100,0.137300,-3.242300])
    input_floats=ffi.new("float[]",[-3.286300,0.387500,-3.487700,0.404300,-3.934100,0.547400,-3.210600])
    translation_floats=ffi.new("float[]",[0,0,20,45,0,0])
    output_floats=ffi.new("float[]",[0,0,0,0,0,0,0])
    translation_dist_unit=ffi.new("char[]","mm".encode("ascii"))
    translation_angle_unit=ffi.new("char[]","degrees".encode("ascii"))
    force_unit_str=ffi.new("char[]","N".encode("ascii"))
    torque_unit_str=ffi.new("char[]","N-m".encode("ascii"))

    lib.SetForceUnits(calibrationData,force_unit_str)
    lib.SetTorqueUnits(calibrationData,torque_unit_str)
    print("Running tool transform with translation: [",end="")
    for i in range(6):
        print(translation_floats[i],end="")
        if (i<5):
            print(",",end="")
    print("]")
    lib.SetToolTransform(calibrationData,translation_floats,translation_dist_unit,translation_angle_unit)
    
    lib.Bias(calibrationData,bias_readings)
    lib.ConvertToFT(calibrationData,input_floats,output_floats)

    print("Bias reading:")
    for i in range(7):
        print(bias_readings[i],end='')
        if (i < 6):
            print(",",end='')
    print("")

    print("Measurement:")
    for i in range(7):
        print(input_floats[i],end='')
        if (i < 6):
            print(",",end='')
    print("")

    print("Result:")
    for i in range(7):
        print(output_floats[i],end='')
        if (i < 6):
            print(",",end='')
    print("")
    
    return 0;

if __name__ == "__main__":

    logging.basicConfig()
    usage_example()
